subprojects/docs/src/docs/userguide/javaPlugin.adoc

```diff
@@ -520,11 +525,18 @@ You can then register it for incremental compilation using the `getSupportedOpti
 </sample>
 ++++
 
-Processors that don't fit these categories will result in full recompilation.
-This includes processors that use `java.io` instead of the `Filer` API and processors that need to read or write resource files.
+Both categories have the following limitations:
+
+* Must not have any side effects except for using the `Filer` and `Messager` APIs.
+For instance, writing files using `java.io` will result in silent failures later on, as these files won't be cleaned up correctly.
+If your processor does this, it cannot be incremental.
+* Must not depend on compiler-specific APIs like `com.sun.source.util.Trees`. Gradle wraps the processing APIs, so attempts to cast to compiler-specific types will fail.
+If your processor does this, it cannot be incremental, unless you have some fallback mechanism.
+* Using `Filer#getResource` currently results in full recompilation. See https://github.com/gradle/gradle/issues/4701[gradle/issues/4701]
+* Using `Filer#createResource` currently results in full recompilation. See https://github.com/gradle/gradle/issues/4702[gradle/issues/4702]
 
 ====== Isolating annotation processors
```

Comment:
Can you think of a way to improve this? It's not clear whether "isolating" is a present participle or an adjective. I'm guessing the latter, but I don't know, and the meaning is quite different in each case.

Reply:
Is it really not clear, even though the section leads with

`Gradle supports incremental compilation for two common categories of annotation processors: "isolating" and "aggregating".`
?

Reply:
I'm convinced I replied to this, so apologies if you end up seeing two similar comments.

This would be fine for users that are reading through the entire section. But for those scanning or skimming the chapter, the headings stand out and it's easy to read as a section on "how to isolate annotation processors".

The more I think about it, the more I'd like to see "Isolating" and "Aggregating" quoted in the headings. What do you think?